### DISCUSSION TITLE
Make a `SessionTeamMiddleware` and `TeamResolver` for easy team feature use

### DISCUSSION THREAD

**Discussion by erikn69** [spatie/laravel-permission#2075] [Ideas]
- #2068
- #1906
- #1998
- #1947
- #1840

```php
// Contract Interface
interface TeamResolver
{
    /**
     * Resolve the Team.
     *
     * @return interger|string|null
     */
    public static function resolve();
}
```
```php
// Kernel Middleware
class SessionTeamMiddleware{
    /**
     * Handle session team_id
     *
     * @param  \Illuminate\Http\Request  $request
     * @param  \Closure  $next
     * @return mixed
     */
    public function handle($request, Closure $next){
        $teamResolver = config('permission.team_resolver');

        if (is_subclass_of($teamResolver , \Contracts\TeamResolver::class)) {
            setPermissionsTeamId(call_user_func([$teamResolver, 'resolve']));
        }

        return $next($request);
    }
}
```
For example here we can have a custom resolver, for cookie, session, token, etc, as default could be session
```php
// TeamResolver example
class TeamResolver implements \Contracts\TeamResolver
{
    /**
     * @return interger|string|null
     */
    public static function resolve()
    {
        return session(PermissionRegistrar::$teamsKey);
    }
}
```
So after login we can set `session(['team_id'=>request()->input('team_id')])`

**CC:** @freekmurze

**Comment by PerryRylance** [comment]
How about a simple `assignRoleOnTeam` which will handle the global?

**Reply by erikn69** [reply]
`assignRoleOnTeam` ?? I think you are mistaking the point of this thread.

**Reply by PerryRylance** [reply]
Sorry if that is the case - It was suggested I come here from #2208

**Reply by erikn69** [reply]
This is about setting team_id on every call after login
>It would be nice if there was an interface that wrapped this in a more elegant way rather than setting a global current team

Maybe the suggestion was about "interface", not about this thread, read again
